proof/refine/AARCH64/TcbAcc_R.thy

```diff
@@ -2043,47 +2851,339 @@ lemma gets_the_exec: "f s \<noteq> None \<Longrightarrow>  (do x \<leftarrow> ge
                    return_def assert_opt_def)
   done
 
+lemma tcbQueueRemove_no_fail:
+  "no_fail (\<lambda>s. tcb_at' tcbPtr s
+                \<and> (\<exists>ts. list_queue_relation ts queue (tcbSchedNexts_of s) (tcbSchedPrevs_of s)
+                        \<and> tcbPtr \<in> set ts)
+                \<and> sym_heap_sched_pointers s \<and> valid_objs' s)
+           (tcbQueueRemove queue tcbPtr)"
+  unfolding tcbQueueRemove_def
+  apply (wpsimp wp: getObject_tcb_wp)
+  apply normalise_obj_at'
+  apply (frule (1) ko_at_valid_objs')
+   apply fastforce
+  apply (clarsimp simp: list_queue_relation_def)
+  apply (prop_tac "tcbQueueHead queue \<noteq> Some tcbPtr \<longrightarrow> tcbSchedPrevs_of s tcbPtr \<noteq> None")
+   apply (rule impI)
+   apply (frule not_head_prev_not_None[where p=tcbPtr])
+      apply (fastforce simp: inQ_def opt_pred_def opt_map_def obj_at'_def)
+     apply (fastforce dest: heap_path_head)
+    apply fastforce
+   apply (fastforce simp: opt_map_def obj_at'_def valid_tcb'_def valid_bound_tcb'_def)
+  by (fastforce dest!: not_last_next_not_None[where p=tcbPtr]
+                 simp: queue_end_valid_def opt_map_def obj_at'_def valid_obj'_def valid_tcb'_def)
+
+crunch (no_fail) no_fail[wp]: removeFromBitmap
+
+crunches removeFromBitmap
+  for ready_queues_relation[wp]: "ready_queues_relation s"
+  and list_queue_relation[wp]:
+   "\<lambda>s'. list_queue_relation ts (P (ksReadyQueues s'))
+                             (tcbSchedNexts_of s') (tcbSchedPrevs_of s')"
+  (simp: bitmap_fun_defs ready_queues_relation_def)
+
+\<comment> \<open>A direct analogue of tcbQueueRemove\<close>
```

Comment:
I'm a bit lost as to what this is for. Maybe the comment could indicate that? Only knowing it's an analogue doesn't explain its introduction.